src/v/storage/segment.cc

```diff
@@ -297,37 +297,29 @@ ss::future<> segment::release_appender(readers_cache* readers_cache) {
 }
 
 void segment::release_appender_in_background(readers_cache* readers_cache) {
-    auto a = std::exchange(_appender, nullptr);
-    auto c = config::shard_local_cfg().release_cache_on_segment_roll()
-               ? std::exchange(_cache, std::nullopt)
-               : std::nullopt;
-    auto i = std::exchange(_compaction_index, std::nullopt);
-    ssx::spawn_with_gate(
-      _gate,
-      [this,
-       readers_cache,
-       a = std::move(a),
-       c = std::move(c),
-       i = std::move(i)]() mutable {
-          return readers_cache
-            ->evict_range(
-              _tracker.get_base_offset(), _tracker.get_dirty_offset())
-            .then(
-              [this, a = std::move(a), c = std::move(c), i = std::move(i)](
-                readers_cache::range_lock_holder readers_cache_lock) mutable {
-                  return ss::do_with(
-                           std::move(readers_cache_lock),
-                           [this](auto&) { return write_lock(); })
-                    .then([this,
-                           a = std::move(a),
-                           c = std::move(c),
-                           i = std::move(i)](ss::rwlock::holder h) mutable {
-                        return do_release_appender(
-                                 std::move(a), std::move(c), std::move(i))
-                          .finally([h = std::move(h)] {});
-                    });
-              });
-      });
+    ssx::spawn_with_gate(_gate, [this, readers_cache]() mutable {
```

Comment:
Does this mean we lose the guarantee upon returning in the general case that _appender is null after calling? If the result of calling this when the gate is closed is that it doesn't close _appender anyway, I wonder if we should just keep the old impl, but explicitly check the gate before doing anything

Reply:
>Does this mean we lose the guarantee upon returning in the general case that _appender is null after calling

FWIW I'm not sure if this property is actually important, but it's a notable change in behavior that seems like it _could_ be relied on elsewhere

Reply:
>Does this mean we lose the guarantee upon returning in the general case that _appender is null after calling?

Yes it does, but if you consider the race laid out in the cover letter, this edge case would previously manifest as a triggered `vassert()` immediately. 

In all other cases, `_appender` will be `nullptr` and closed correctly. It is lucky that the general case has never hit this.

If we want to enforce that `_appender` must ` == nullptr` post `release_appender()`, which I think is a very fair contract, we could always move the gate holder call to within `release_appender()` instead?

>If the result of calling this when the gate is closed is that it doesn't close _appender anyway, I wonder if we should just keep the old impl, but explicitly check the gate before doing anything

I do not see how the outcome is different if we early return upon encountering a closed gate, versus silently catch an exception by entering a closed gate with `ssx::spawn_with_gate()`?

Reply:
I mean that if the gate is open, aren't we losing the guarantee that the appender exchange happens during this call? Agreed that we don't have such a guarantee in the general case, but I just want to make sure we're not regressing the common case

Reply:
Ah, I see now. 

The other change I think we could make if we didn't want to lose this guarantee is to flip the order of `co_await _gate.close()` and `co_await write_lock()` in `segment::close()`, though that change in behavior could introduce other side effects as well.

What do you think?

Reply:
Thanks!